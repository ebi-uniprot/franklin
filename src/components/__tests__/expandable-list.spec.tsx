import { Fragment } from 'react';
import { screen, render, fireEvent } from '@testing-library/react';

import ExpandableList, { ExpandableMessage } from '../expandable-list';

const descriptionString = 'foo';

describe('ExpandableMessage', () => {
  let setExpanded: () => void;
  beforeEach(() => {
    setExpanded = jest.fn();
  });

  test('should render', () => {
    const { asFragment } = render(
      <ExpandableMessage
        expanded={false}
        setExpanded={setExpanded}
        descriptionString={descriptionString}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should call setExpanded when button is clicked and button text be "More descriptionString"', () => {
    render(
      <ExpandableMessage
        expanded={false}
        setExpanded={setExpanded}
        descriptionString={descriptionString}
      />
    );
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(setExpanded).toHaveBeenCalled();
    expect(button.textContent).toEqual(`More ${descriptionString}`);
  });

  test('button text should be "Less descriptionString"', () => {
    render(
      <ExpandableMessage
        expanded
        setExpanded={setExpanded}
        descriptionString={descriptionString}
      />
    );
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(setExpanded).toHaveBeenCalled();
    expect(button.textContent).toEqual(`Less ${descriptionString}`);
  });

  test('button text should include the number of hidden items', () => {
    render(
      <ExpandableMessage
        setExpanded={setExpanded}
        descriptionString={descriptionString}
        nHiddenItems={4}
      />
    );
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(setExpanded).toHaveBeenCalled();
    expect(button.textContent).toEqual(`4 more ${descriptionString}`);
  });
});

describe('ExpandableList', () => {
  const numberItems = 10;
  const numberCollapsedItems = 5;
  const items = Array.from({ length: numberItems }, (_, index) => (
    <Fragment key={index}>some content {index}</Fragment>
  ));

  test('should render', () => {
    const { asFragment } = render(
      <ExpandableList
        numberCollapsedItems={numberCollapsedItems}
        descriptionString={descriptionString}
      >
        {items}
      </ExpandableList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should not render when no children', () => {
    const { container } = render(
      <ExpandableList
        numberCollapsedItems={numberCollapsedItems}
        descriptionString={descriptionString}
      />
    );
    expect(container).toBeEmptyDOMElement();
  });

  test('should not render when empty list as children', () => {
    const { container } = render(
      <ExpandableList
        numberCollapsedItems={numberCollapsedItems}
        descriptionString={descriptionString}
      >
        {[]}
      </ExpandableList>
    );
    expect(container).toBeEmptyDOMElement();
  });

  test('should not show more button if numberItems <= numberCollapsedItems + 1', () => {
    render(
      <ExpandableList
        numberCollapsedItems={numberCollapsedItems}
        descriptionString={descriptionString}
      >
        {items.slice(0, numberCollapsedItems + 1)}
      </ExpandableList>
    );
    expect(screen.queryByTestId('expandable-message')).toBeNull();
  });

  test('should not show more button if numberItems > numberCollapsedItems + 1', () => {
    render(
      <ExpandableList
        numberCollapsedItems={numberCollapsedItems}
        descriptionString={descriptionString}
      >
        {items}
      </ExpandableList>
    );
    expect(screen.getByTestId('expandable-message')).toBeTruthy();
  });

  test('should initially have numberCollapsedItems and then total number of items after More button click', () => {
    const { container } = render(
      <ExpandableList
        numberCollapsedItems={numberCollapsedItems}
        descriptionString={descriptionString}
      >
        {items}
      </ExpandableList>
    );
    let listItems = container.querySelectorAll('li');
    expect(listItems).toHaveLength(numberCollapsedItems + 1);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    listItems = container.querySelectorAll('li');
    expect(listItems).toHaveLength(numberItems + 1);
  });
});
