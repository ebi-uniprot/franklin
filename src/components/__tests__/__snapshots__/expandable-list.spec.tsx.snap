// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableList should render 1`] = `
<DocumentFragment>
  <ul
    class="expandable-list no-bullet"
  >
    <li>
      some content 0
    </li>
    <li>
      some content 1
    </li>
    <li>
      some content 2
    </li>
    <li>
      some content 3
    </li>
    <li>
      some content 4
    </li>
    <li>
      <button
        class="button tertiary expandable-list__action"
        data-testid="expandable-message"
        type="button"
      >
        More foo
      </button>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ExpandableMessage should render 1`] = `
<DocumentFragment>
  <button
    class="button tertiary expandable-list__action"
    data-testid="expandable-message"
    type="button"
  >
    More foo
  </button>
</DocumentFragment>
`;
